File: Python/Sort and Search/BinarySearch.py
# Implementation of a Binary Search

from nose.tools import assert_equal

# Using recursion
def binary_search_recursion(arr, target):
	if len(arr) > 0:
		mid = len(arr)//2
		print(arr[mid])
		if arr[mid] == target:
			return True
		if target < arr[mid]:
			return binary_search_recursion(arr[:mid], target)
		else:
			return binary_search_recursion(arr[mid+1:], target)
	else:
		return False

# Using iteration
def binary_search_iteration(arr, target):
	first = 0
	last = len(arr) - 1
	while first <= last:
		mid = first + last
		if arr[mid] == target:
			return True
		else:
			if target < arr[mid]:
				last = mid - 1
			else:
				first = mid + 1
	return False

# print(binary_search([1, 2, 3, 4, 5, 6, 7, 8], 1))

class Binary_Search(object):
	def test(self, func):
		assert_equal(func([1, 2, 3, 4, 5, 6, 7, 8], 1), True)
		assert_equal(func([1, 2, 3, 4, 5, 6, 7, 8], 4), True)
		assert_equal(func([1, 2, 3, 4, 5, 6, 7, 8], 0), False)
		assert_equal(func([1, 2, 3, 4, 5, 6, 7, 8], 9), False)
		assert_equal(func([1, 2, 3, 4, 6, 7, 8], 5), False)
		print('TESTS PASSED')

t = Binary_Search()
t.test(binary_search_recursion)
t.test(binary_search_iteration)

File: Sort and Search/BinarySearch.js
function binarySearch(arr, n){
	var first = 0;
	var last = arr.length;
	while(first <= last){
		var mid = Math.floor((first+last)/2);
		if(n == arr[mid]){
			return true;
		} else if(n < arr[mid]){
			last = mid-1;
		} else{
			first = mid+1;
		}
	}
	return false;
}

function recBinSearch(arr, n){
	mid = Math.floor(arr.length/2);
	if(arr.length < 1){
		return false;
	}
	if(arr[mid] === n){
		return true;
	}
	if(n < arr[mid]){
		return recBinSearch(arr.slice(0, mid), n);
	} else{
		return recBinSearch(arr.slice(mid+1, arr.length), n);
	}
}

console.log(binarySearch([1, 2, 4, 5, 6, 7], 1));
console.log(binarySearch([1, 2, 4, 5, 6, 7], 2));
console.log(binarySearch([1, 2, 4, 5, 6, 7], 4));
console.log(binarySearch([1, 2, 4, 5, 6, 7], 5));
console.log(binarySearch([1, 2, 4, 5, 6, 7], 6));
console.log(binarySearch([1, 2, 4, 5, 6, 7], 7));
console.log(binarySearch([1, 2, 4, 5, 6, 7], 0));
console.log(binarySearch([1, 2, 4, 5, 6, 7], 8));
console.log(binarySearch([1, 2, 5, 6, 7], 18));

console.log(" ");

console.log(recBinSearch([1, 2, 4, 5, 6, 7], 1));
console.log(recBinSearch([1, 2, 4, 5, 6, 7], 2));
console.log(recBinSearch([1, 2, 4, 5, 6, 7], 4));
console.log(recBinSearch([1, 2, 4, 5, 6, 7], 5));
console.log(recBinSearch([1, 2, 4, 5, 6, 7], 6));
console.log(recBinSearch([1, 2, 4, 5, 6, 7], 7));
console.log(recBinSearch([1, 2, 4, 5, 6, 7], 0));
console.log(recBinSearch([1, 2, 4, 5, 6, 7], 8));
console.log(recBinSearch([1, 2, 5, 6, 7], 18));